### PR TITLE
Include type restriction on getters in JSON/mapping

### PR DESCRIPTION
**Pull request by Daniel-Worrall**
Addresses #5920

Was unsure of a better way to spec this but I'm very aware of how dirty it is and how it relies on macro code more than anything else.

### CODE REVIEW

File: src/json/mapping.cr [hunk 1]
```diff
@@ -80,7 +80,7 @@ module JSON
       {% end %}
 
       {% if value[:getter] == nil ? true : value[:getter] %}
-        def {{key.id}}
+        def {{key.id}} : {{value[:type]}} {{ (value[:nilable] ? "?" : "").id }}
```
Comment (Sija) [inline src/json/mapping.cr]: `?` should come directly after the type (`Bool?` not `Bool ?`)

File: spec/std/json/mapping_spec.cr [hunk 2]
```diff
@@ -507,6 +507,14 @@ describe "JSON mapping" do
       json.bar?.should be_false
     end
 
+    it "defines query getter with class restriction" do
```
[the remark this answers is not in the document]
Comment (Daniel-Worrall) [reply spec/std/json/mapping_spec.cr]: It being optional doesn't mean we shouldn't be testing the optional case of supplying it.

File: spec/std/json/mapping_spec.cr [hunk 3]
```diff
@@ -507,6 +507,14 @@ describe "JSON mapping" do
...
+      {% begin %}
+      {% methods = JSONWithQueryAttributes.methods %}
+      {{methods.find{ |m| m.name == "foo?"}.return_type }}.should eq(Bool)
```
[the remark this answers is not in the document]
Comment (Daniel-Worrall) [reply spec/std/json/mapping_spec.cr]: This code does not work.

File: src/json/mapping.cr (hunk 1 above)
Comment (Daniel-Worrall) [reply src/json/mapping.cr]: This is exactly as it is in the setter above. Should both be changed?

Comment (straight-shoota) [reply src/json/mapping.cr]: It parses with whitespace, but it would be better to remove it.

File: spec/std/json/mapping_spec.cr [hunk 4]
```diff
@@ -507,6 +507,14 @@ describe "JSON mapping" do
       json.bar?.should be_false
     end
 
+    it "defines query getter with class restriction" do
+      {% begin %}
+        {% methods = JSONWithQueryAttributes.methods %}
+        {{methods.find(&.name.==("foo?")).return_type}}.should eq(Bool)
```
Comment (Sija) [inline spec/std/json/mapping_spec.cr]: Add the whitespaces, please!

**Comment by Daniel-Worrall** [comment]
This should be good to go assuming we approve the spec being in place

**Pull request merged**